openprescribing/frontend/views/spending_utils.py

```diff
@@ -14,228 +20,351 @@
-            _date_as_str(start_date), _date_as_str(last_prescribing_date)),
-        # We discount by an additional factor of 100 to convert the figures
-        # from pence to pounds
-        'discount_factor': (100 - NATIONAL_AVERAGE_DISCOUNT_PERCENTAGE) / (100*100)
+
+
+def _get_concession_price_matrices(min_date, max_date):
+    """
+    Return details of NCSO price concessions in the given date range
+    (inclusive)
+
+    Returns a ConcessionPriceMatrices object with the following attributes:
+
+        bnf_code_offsets: Maps BNF codes which have a price concession to their
+                          row offset in the matrices
+
+            date_offsets: Maps date strings to their column offset in the
+                          matrices
+
+           tariff_prices: Matrix giving, for each BNF code and date, the
+                          standard Drug Tariff price for that presentation (in
+                          pounds-per-unit)
+
+         price_increases: Matrix giving, for each BNF code and date, the price
+                          increase due to concessions (in pounds-per-unit)
+
+    """
+    # Fetch list of concessions and associated tariff prices from Postgres
+    concessions = _get_concession_prices(min_date, max_date)
+    # Construct the BNF code and date indices we need to store these prices in
+    # matrix form
+    date_offsets = {
+        str(date): i
+        for (i, date) in enumerate(_get_dates_in_range(min_date, max_date))
     }
-    return sql, params
+    bnf_codes = {concession[1] for concession in concessions}
+    bnf_code_offsets = {bnf_code: i for (i, bnf_code) in enumerate(bnf_codes)}
+    # Construct the matrices we need
+    shape = (len(bnf_code_offsets), len(date_offsets))
+    tariff_prices = numpy.zeros(shape, dtype=numpy.float_)
+    price_increases = numpy.zeros(shape, dtype=numpy.float_)
+    # Loop over the concessions and write them into the matrices
+    for date, bnf_code, tariff_price, price_increase, quantity_per_pack in concessions:
+        index = (bnf_code_offsets[bnf_code], date_offsets[str(date)])
+        # Convert prices from per-pack into per-unit
+        tariff_price = tariff_price / quantity_per_pack
+        price_increase = price_increase / quantity_per_pack
+        # Price concessions are defined at the product-pack level but we only
+        # have prescribing data at product level. Occasionally there are
+        # multiple simultaneous concessions for a given product with different
+        # implied prices-per-unit for different pack sizes. As we can't know
+        # from our data which pack size was dispensed we just use the highest
+        # per-unit price.
+        if price_increase > price_increases[index]:
+            price_increases[index] = price_increase
+            tariff_prices[index] = tariff_price
+    # Convert prices from pence to pounds and apply the national average
+    # discount to get a better approximation of the actual price paid
```

Comment:
I don't understand this.

Reply:
Does the comment on the NATIONAL_AVERAGE_DISCOUNT constant help at all?
https://github.com/ebmdatalab/openprescribing/blob/a30136a1936ddc3390d1c64ac0bf94cbab680a7f/openprescribing/frontend/views/spending_utils.py#L17-L20

Reply:
No, what I don't understand is why converting from pence to pounds gets a better approximation.

But now I come back to it, it's applying the discount that gets a better approximation, and converting from pence to pounds is unrelated.  (Right?)

Is it worth splitting the calculation into two steps for clarity?

Reply:
Ah yes, I see now that the comment is very unhelpful. I've reworded it to be less misleading. I'd rather not do two big matrix multiplications where one would do unless it's a significant win for clarity.

Reply:
Good stuff.